### PR TITLE
fix: fix bugs for rebuilding and snapshot purge then add more test steps

### PR DESCRIPTION
**Pull request by shuo-wu**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#7199

### CODE REVIEW

**Review by derekbit — APPROVED** [review]

**Comment by derekbit** [comment]
LGTM

**Comment by derekbit** [comment]
@shuo-wu Is it ready for review?

**Review by derekbit — APPROVED** [review]
LGTM

File: pkg/spdk_test.go [hunk 1]
```diff
@@ -789,51 +790,232 @@ func (s *TestSuite) TestSPDKMultipleThreadSnapshot(c *C) {
...
+			revertSnapshot(c, spdkCli, snapshotName15, volumeName, engineName, replicaAddressMap)
+
+			// Test online rebuilding twice
+
+			// Write some data to the head before the 1st rebuilding
+			offsetInMB = 6 * dataCountInMB
+			_, err = ne.Execute(nil, "dd", []string{"if=/dev/urandom", fmt.Sprintf("of=%s", endpoint), "bs=1M", fmt.Sprintf("count=%d", dataCountInMB), fmt.Sprintf("seek=%d", offsetInMB), "status=none"}, defaultTestExecuteTimeout)
 			c.Assert(err, IsNil)
-			err = spdkCli.EngineReplicaDelete(engineName, replicaName2, net.JoinHostPort(ip, strconv.Itoa(int(replica2.PortStart))))
+			cksumBeforeRebuild11, err := util.GetFileChunkChecksum(endpoint, offsetInMB*helpertypes.MiB, dataCountInMB*helpertypes.MiB)
```
Comment (c3y1huang) [inline pkg/spdk_test.go:802]: > cksumBeforeRebuild11

What is the `11`?

File: pkg/spdk/util.go [hunk 2]
```diff
@@ -126,9 +126,10 @@ func CleanupLvolTree(spdkClient *spdkclient.Client, rootLvolName string, bdevLvo
 		}
 	}
 	for idx := len(queue) - 1; idx >= 0; idx-- {
+		// This may fail since there may be a rebuilding failed replicas on the same host that leaves an orphan rebuilding lvol as a child of a snapshot lvol.
+		// Then this snapshot lvol would have multiple children then cannot be deleted.
 		if _, err := spdkClient.BdevLvolDelete(queue[idx].UUID); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
-			return err
+			log.WithError(err).Errorf("failed to delete lvol %v(%v) from the lvol tree with root %v(%s), this lvol may accidentally have some leftover orphans chidlren %+v, will continue", queue[idx].Aliases[0], queue[idx].UUID, bdevLvolMap[rootLvolName].Aliases[0], bdevLvolMap[rootLvolName].UUID, queue[idx].DriverSpecific.Lvol.Clones)
```
Comment (c3y1huang) [inline pkg/spdk/util.go:132]:
```suggestion
			log.WithError(err).Errorf("Failed to delete lvol %v(%v) from the lvol tree with root %v(%s), this lvol may accidentally have some leftover orphans children %+v, will continue", queue[idx].Aliases[0], queue[idx].UUID, bdevLvolMap[rootLvolName].Aliases[0], bdevLvolMap[rootLvolName].UUID, queue[idx].DriverSpecific.Lvol.Clones)
```

File: pkg/spdk_test.go (hunk 1 above)
Comment (shuo-wu) [reply pkg/spdk_test.go:802]: branch 1 - rebuilding snapshot 1

**Review by derekbit — APPROVED** [review]

**Comment by derekbit** [comment]
Merge first. @shuo-wu will continue investigating the errors.

**Pull request merged**